### ISSUE TITLE
[Exit code -6] malloc_error_break

### ISSUE THREAD

**Issue by biomendi** [voutcn/megahit#160]
I am running MEGAHIT as a part of the pipeline NORGAL (https://bitbucket.org/kosaidtu/norgal) to extract mitochondrial genomes from NGS data. My data consists of two fastq.gz that are ~500MB each. My server is a Mac Pro with 24 CPU threads and 64GB of RAM (which is not being used by anything else - apart from the system). I encountered the following error which I haven't seen mentioned by anyone yet. I would really appreciate any help :)

```
MEGAHIT v1.1.2
--- [Wed Mar 21 14:40:36 2018] Start assembly. Number of CPU threads 20 ---
--- [Wed Mar 21 14:40:36 2018] Available memory: 68719476736, used: 61847529062
--- [Wed Mar 21 14:40:36 2018] Converting reads to binaries ---
/anaconda2/bin/megahit_asm_core buildlib assemblies/initial/tmp/reads.lib assemblies/initial/tmp/reads.lib
    [read_lib_functions-inl.h  : 209]     Lib 0 (reads/trim1.fq,reads/trim2.fq): pe, 15075004 reads, 100 max length
    [utils.h                   : 126]     Real: 25.7118 user: 13.9374   sys: 6.1822     maxrss: 204058624
--- [Wed Mar 21 14:41:02 2018] k list: 21,49,77,105 ---
--- [Wed Mar 21 14:41:02 2018] Extracting solid (k+1)-mers for k = 21 ---
cmd: /anaconda2/bin/megahit_sdbg_build count -k 21 -m 2 --host_mem 61847529062 --mem_flag 0 --gpu_mem 0 --output_prefix assemblies/initial/tmp/k21/21 --num_cpu_threads 20 --num_output_threads 6 --read_lib_file assemblies/initial/tmp
    [sdbg_builder.cpp          : 112]     Host memory to be used: 61847529062
    [sdbg_builder.cpp          : 113]     Number CPU threads: 20
    [cx1.h                     : 450]     Preparing data...
    [read_lib_functions-inl.h  : 256]     Before reading, sizeof seq_package: 493112708
    [read_lib_functions-inl.h  : 260]     After reading, sizeof seq_package: 493112708
    [cx1_kmer_count.cpp        : 136]     15075004 reads, 100 max read length
    [cx1.h                     : 457]     Preparing data... Done. Time elapsed: 3.6311
    [cx1.h                     : 464]     Preparing partitions and initialing global data...
    [cx1_kmer_count.cpp        : 227]     2 words per substring, 2 words per edge
    [cx1.h                     : 171]     Adjusting memory layout: max_lv1_items=18480426, num_sorting_items=1296703, mem_sorting_items=31120872, mem_avail=105042576
    [cx1_kmer_count.cpp        : 356]     18480426, 1296703 105042576 61259147922
    [cx1_kmer_count.cpp        : 363]     Memory for reads: 504753756
    [cx1_kmer_count.cpp        : 364]     max # lv.1 items = 18480426
    [cx1.h                     : 480]     Preparing partitions and initialing global data... Done. Time elapsed: 1.6466
    [cx1.h                     : 486]     Start main loop...
    [cx1.h                     : 515]     Lv1 scanning from bucket 0 to 4
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3926

    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.3867
    [cx1.h                     : 515]     Lv1 scanning from bucket 4 to 9
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3777
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1817
    [cx1.h                     : 515]     Lv1 scanning from bucket 9 to 14
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3779
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1408
    [cx1.h                     : 515]     Lv1 scanning from bucket 14 to 20
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3775
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1382
    [cx1.h                     : 515]     Lv1 scanning from bucket 20 to 26
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3788
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1355
    [cx1.h                     : 515]     Lv1 scanning from bucket 26 to 32
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3845
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1328
    [cx1.h                     : 515]     Lv1 scanning from bucket 32 to 39
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3808
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1502
    [cx1.h                     : 515]     Lv1 scanning from bucket 39 to 47
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3776
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1125
    [cx1.h                     : 515]     Lv1 scanning from bucket 47 to 56
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3880
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1175
    [cx1.h                     : 515]     Lv1 scanning from bucket 56 to 66
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3882

    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.1043
    [cx1.h                     : 515]     Lv1 scanning from bucket 66 to 77
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3807
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.0945
    [cx1.h                     : 515]     Lv1 scanning from bucket 77 to 90
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3824
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.0920
    [cx1.h                     : 515]     Lv1 scanning from bucket 90 to 104
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3805
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.0850
    [cx1.h                     : 515]     Lv1 scanning from bucket 104 to 119
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3994
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.0861
    [cx1.h                     : 515]     Lv1 scanning from bucket 119 to 135
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3808
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.0827
    [cx1.h                     : 515]     Lv1 scanning from bucket 135 to 152
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3819
    [cx1.h                     : 594]     Lv1 fetching & sorting done. Time elapsed: 0.0813
    [cx1.h                     : 515]     Lv1 scanning from bucket 152 to 170
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 1.3830
megahit_sdbg_build(3020,0x7fffaae4e3c0) malloc: *** error for object 0x7fa579c03120: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
Error occurs when running "sdbg_builder count/read2sdbg", please refer to assemblies/initial/log for detail
[Exit code -6] 

```
EDIT: I have encountered a very similar error when I tried to perform the assembly on a reduced version of my data ~ 1MB (I extracted only those reads mapping to reference mitogenomes). The error that I get this time is the following (top lines skipped for convenience):

```
    [cx1.h                     : 515]     Lv1 scanning from bucket 19022 to 65536
    [cx1.h                     : 528]     Lv1 scanning done. Large diff: 0. Time elapsed: 0.0095
megahit_sdbg_build(3998,0x7fffaae4e3c0) malloc: *** error for object 0x7f817f500240: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug

Error occurs when running "builder build" for k = 29; please refer to deduped_ALLreads_mapped_to_mergedRef.fq.megahit_asm/log for detail
[Exit code -6]
```

**Comment by voutcn** [comment]
From the log, it's hard to tell where it goes wrong. Could you provide the subset of reads to me for debugging? Thanks.

**Comment by biomendi** [comment]
Thanks for your reply. I have been able to replicate the error using a much smaller number of reads ~ 1.2MB (fastq format can't be attached here, but I could send it to you by email). 

Moreover, I have noticed something interesting: when using the previous data (or any data), MEGAHIT will more likely output an error when a large number of CPUs (20-24) are used. For example, the following command will give an error most of the times:

`megahit -t 24 -r 'deduped_ALLreads_mapped_to_mergedRef.fq' --min-count '2' --k-min '21' --k-max '141' --k-step '12' --min-contig-len '150'  --bubble-level '2' --merge-level '20,0.95' --prune-level '2' --prune-depth '2' --low-local-ratio '0.2'   && cat megahit_out/log`

However, the next command will run sucessfully:

`megahit -t 2 -r 'deduped_ALLreads_mapped_to_mergedRef.fq' --min-count '2' --k-min '21' --k-max '141' --k-step '12' --min-contig-len '150'  --bubble-level '2' --merge-level '20,0.95' --prune-level '2' --prune-depth '2' --low-local-ratio '0.2'   && cat megahit_out/log`

**Comment by voutcn** [comment]
I just tried your data but cannot reproduce the error (perhaps because my PC has only 8 CPU threads and it's a Linux PC). I am not sure whether this error only happens on MacOS. Will revisit this issue later.

**Comment by lfdelzam** [comment]
I am trying to figure out what Exit code 255 and Exit code -6 mean. Most exit codes are specific for the program that gives them. Where could I find this information?
Thanks,

**Comment by voutcn** [comment]
This should have been fixed in 1.2.x versions. Closing this issue.

@lfdelzam For the meaning of exit code you can refer to https://www.tldp.org/LDP/abs/html/exitcodes.html
However, the purpose of presenting exit codes in Megahit logs is for developers' debugging only. If you have trouble with it, please try the [latest version](https://github.com/voutcn/megahit/releases/tag/v1.2.5-beta). If it doesn't work, just file another issue.
